msvc.8
.TH msvc 8
.SH NAME
msvc \- control minit
.SH SYNOPSIS
.B msvc
[
.B \-[uodpchaitko]
] [
.B \-P
.I pid
]
.I service
[\fI...\fR]
.SH DESCRIPTION
.B msvc
is the management interface to minit.
.I service
is the service directory name relative to /etc/minit.  Starting with
minit 0.9 you can also include /etc/minit/ in the service name.
.SH OPTIONS
If no option is given,
.B msvc
will just print a small diagnostic message to stdout, saying if the
service is up, down or finished, which PID it has if it is up, and for
how long it has been in this state.
.TP 5
.B \-u
Up.
If the service is not running, start it.
If the service stops, restart it.
.TP
.B \-o
Once.
If the service is not running, start it.
If the service stops, do not restart it.
.TP
.B \-d
Down.
If the service is running, send it a TERM signal and then a CONT signal.
After it stops, do not restart it.
.TP
.B \-p
Pause.
Send the service a STOP signal.
.TP
.B \-c
Continue.
Send the service a CONT signal.
.TP
.B \-h
Hangup.
Send the service a HUP signal.
.TP
.B \-a
Alarm.
Send the service an ALRM signal.
.TP
.B \-i
Interrupt.
Send the service an INT signal.
.TP
.B \-t
Terminate.
Send the service a TERM signal.
.TP
.B \-k
Terminate.
Send the service a KILL signal.
.TP
.B \-P \fIpid\fR
Set PID.
Tell minit the PID of the service is really \fIpid\fR.  This is useful
for services that fork themselves in the background but put their real
PID in a file, typically called /var/run/\fIservice\fR.pid.  Used by
\fBpidfilehack\fR.
.TP
.B \-D \fIservice\fR
Print dependencies.
This will print all the names of all the services that were started
because this services depended on them.  Please note that this is not
done recursively (i.e. if default depends on qmail and qmail depends on
log, this will print qmail, not qmail/log.  But msvc -D qmail will print
qmail/log).
.TP
.B \-H
Print history.
This will print the names of the ten least recently spawned processes.
This is useful if you see a process looping (initialization fails and
minit is restarting it all the time).
.SH "RETURN CODES"
Generally, msvc return zero if everything is OK or 1 on error (could not
open /etc/minit/in or /etc/minit/out or there is no process with the
given name).  In diagnostic mode, it will exit 0 if the service is up, 2
if it is down or 3 if it is finished.
.SH "SEE ALSO"
pidfilehack(8), svc(8)


CHANGES
0.10:
  add sample script for /etc/minit/ctrlaltdel/run as
    control/ctrlaltdel-run (Florian Westphal)
  add killall5, needs /proc on Linux though (Florian Westphal)
  add shutdown (Bernd Wachter)
  add -Wall and -W and kill resulting compiler warnings
  several fixes (Bernd Wachter)
  add update support - tell minit to execve something else, a newer
    version of itself, preferably (Florian Westphal)
  also add minit-update program to download state from old minit and
    upload state to new minit after update (Florian Westphal)
  #include minit.h and use MINITROOT consistently (Lordy)
  Don't assume MINITROOT is 10 bytes long (Erich Schubert)
  add man pages!! (Erich Schubert)
  add a way to ask minit for dependencies
  msvc -H will now dump the ten least recently spawned processes.
    This is useful to see which process is looping.
  fix bug in msvc killing process group if readpid failed
    (Erich Schubert)
  You can now blacklist services temporarily by passing -servicename on
    the kernel command line.  The first time they are to be started will
    then be skipped.  Use this, for example, to not start the dhcp
    client when your notebook is not plugged in a network.
  add serdo (execute batch, no shell features whatsoever)
  make "msvc servicename" output easier to parse if not outputting to a tty
  require libowfat instead of shipping half of it ourselves
  replace malloc with alloca; minit is now 6644 bytes on i386.
  add usage for serdo (Nikola Vladov)
  fix msvc -d followed by msvc -h (would send SIGHUP to pid 1, i.e. kill
    minit; found by Paul Fox)
  added small special-purpose malloc using a static buffer before
    calling the real malloc (Nikola Vladov)

0.9.1:
  fix embarassing typo in msvc (Gelu G. Lupas)

0.9:
  For stopped services, minit did not record the time when it finished,
    but when it was started.
  add hard-reboot from Tommi Virtanen.  He also asked me to include an
    explicit license statement and sent a patch to make debian packaging
    easier (deb files can't include FIFOs).
  Olaf: add write_proc.
  Olaf: lines in depends can now be commented out with a #
  Olaf: add a method to clear dead services (mark as terminated)
  minit now checks whether a PID actually exists before accepting msvc -P
    this removes a race condition when the forked service terminates
    before pidfilehack notifies minit of the PID
  msvc now accepts more than one service after -o, -d, ...
  msvc now accepts /etc/minit/sshd instead of sshd as service name
  better error handling in msvc (more and better error messages)
  minit now has a heuristic to detect time zone changes and still
    have msvc report the correct elapsed time since process start/end
    (don't look at the source code, it's ultra-kludgey)
  add msvc -g (get PID for shell scripts)

0.8:
  call waitpid repeatedly until it returns "no children".
  This reaps zombies faster.

0.7:
  found and fixed the bug that made the logging support not work when
    minit was running as PID 1.
  saw that the shutdown from sysvinit actually does what it's supposed
    to do, even with minit, if one uses the -n mode.  It should be quite
    easy to write a nice and clean shutdown for minit now.
  if a service depends on a service that already finished, don't start
    it again.

0.6.1:
  fixed msvc and minit handling of msvc -P.
  pidfilehack has now actually been tested and works with ssh.

0.6:
  add tty and session leadership magic and I actually got getty to work
    as init on my laptop, i.e. in production use!
  add msvc man page.

0.5.1:
  cut&paste error, set wrong variable to 0 for keyboard request.
    Thanks, Schulti.

0.5:
  add "log" subdirectory (see README).
  get rid of stdio and *printf and switch to routines from libowfat.

0.4:
  add pidfilehack.  It will take the first argument as filename for the
  PID file (i.e. /var/run/sshd.pid), the second argument as service name
  for minit (i.e. ssh) and the rest of the command line is run without
  path searching in a child process.  The parent process will then
  try to open the pidfile several times (for up to 30 seconds), read the
  PID off it and run msvc to tell minit the PID of the service.  minit
  can then track sshd and restart it if necessary.  This gross hack does
  not work for daemons that do not write a PID file, but there are not
  many of those, fortunately.


msvc.c
#include <fcntl.h>
#include <sys/file.h>
#include <signal.h>
#include <stdio.h>
#include <string.h>
#include <stdlib.h>
#include <unistd.h>
#include "str.h"
#include "fmt.h"
#include "buffer.h"
#define NOVARS
#include "minit.h"
#include <errmsg.h>
#include <errno.h>

static int infd,outfd;

static char buf[1500];

void addservice(char* service) {
  char* x;
  if (str_start(service,MINITROOT "/"))
    service+=sizeof(MINITROOT "/") -1;
  x=service+str_len(service)-1;
  while (x>service && *x=='/') { *x=0; --x; }
  strncpy(buf+1,service,1400);
  buf[1400]=0;
}

int addreadwrite(char* service) {
  addservice(service);
  write(infd,buf,str_len(buf));
  return read(outfd,buf,1500);
}

/* return PID, 0 if error */
pid_t __readpid(char *service) {
  int len;
  buf[0]='p';
  len=addreadwrite(service);
  if (len<0) return 0;
  buf[len]=0;
  return atoi(buf);
}

/* return nonzero if error */
int respawn(char *service,int yesno) {
  int len;
  buf[0]=yesno?'R':'r';
  len=addreadwrite(service);
  return (len!=1 || buf[0]=='0');
}

/* return nonzero if error */
int setpid(char *service, pid_t pid) {
  char *tmp;
  int len;
  buf[0]='P';
  addservice(service);
  tmp=buf+str_len(buf)+1;
  tmp[fmt_ulong(tmp,pid)]=0;
  write(infd,buf,str_len(buf)+str_len(tmp)+2);
  len=read(outfd,buf,1500);
  return (len!=1 || buf[0]=='0');
}

/* return nonzero if error */
int check_remove(char *service) {
  int len;
  buf[0]='C';
  len=addreadwrite(service);
  return (len!=1 || buf[0]=='0');
}

/* return nonzero if error */
int startservice(char *service) {
  int len;
  buf[0]='s';
  len=addreadwrite(service);
  return (len!=1 || buf[0]=='0');
}

/* return uptime, 0 if error */
unsigned long uptime(char *service) {
  int len;
  buf[0]='u';
  len=addreadwrite(service);
  if (len<0) return 0;
  buf[len]=0;
  return atoi(buf);
}

void dumphistory() {
  char tmp[16384];
  int i,j;
  char first,last;
  first=1; last='x';
  write(infd,"h",1);
  for (;;) {
    int prev,done;
    j=read(outfd,tmp,sizeof(tmp));
    if (j<1) break;
    done=i=0;
    if (first) {
      if (tmp[0]=='0') {
	carp("minit compiled without history support.");
	return;
      }
      i+=2;
    } else {
      if (!tmp[0] && last=='\n') break;
    }
    prev=i;
    for (; i<j; ++i)
      if (!tmp[i]) {
	tmp[i]=done?0:'\n';
	if (i<j && !tmp[i+1]) { done=1; --j; }
      }
    if (first)
      write(1,tmp+2,j-2);
    else
      write(1,tmp,j);
    if (done) break;
    last=tmp[j-1];
    first=0;
  }
}

void dumpdependencies(char* service) {
  char tmp[16384];
  int i,j;
  char first,last;
  buf[0]='d';
  addservice(service);
  write(infd,buf,str_len(buf));
  first=1; last='x';
  for (;;) {
    int prev,done;
    j=read(outfd,tmp,sizeof(tmp));
    if (j<1) break;
    done=i=0;
    if (first) {
      if (tmp[0]=='0') {
	carp(service,": no such service.");
	return;
      }
      i+=2;
    } else {
      if (!tmp[0] && last=='\n') break;
    }
    prev=i;
    for (; i<j; ++i)
      if (!tmp[i]) {
	tmp[i]=done?0:'\n';
	if (i<j && !tmp[i+1]) { done=1; --j; }
      }
    if (first)
      write(1,tmp+2,j-2);
    else
      write(1,tmp,j);
    if (done) break;
    last=tmp[j-1];
    first=0;
  }
}

int main(int argc,char *argv[]) {
  if (argc<2) {
    msg(
	"usage: msvc -[uodpchaitkogC] service\n"
	"       msvc -Ppid service\n"
	" -u\tup; start service with respawn\n"
	" -o\tonce; start service without respawn\n"
	" -d\tdown; disable respawn, stop service\n"
	" -p\tpause; send SIGSTOP\n"
	" -c\tcontinue; send SIGCONT\n"
	" -h\thangup; send SIGHUP\n"
	" -a\talarm; send SIGALRM\n"
	" -i\tintr; send SIGINT\n"
	" -t\tterminate; send SIGTERM\n"
	" -k\tkill; send SIGKILL\n"
	" -g\tget; output just the PID\n"
	" -Ppid\tset PID of service (for pidfilehack)\n"
	" -D service\tprint services started as dependency\n"
	" -H\tprint last n respawned services\n"
	" -C\tClear; remove service form active list\n");
    return 0;
  }
  errmsg_iam("msvc");
  infd=open(MINITROOT "/in",O_WRONLY);
  outfd=open(MINITROOT "/out",O_RDONLY);
  if (infd>=0) {
    while (lockf(infd,F_LOCK,1)) {
      carp("could not acquire lock!");
      sleep(1);
    }
    if (argc==2 && argv[1][1]!='H') {
      pid_t pid=__readpid(argv[1]);
      if (buf[0]!='0') {
	unsigned long len;
	unsigned long ut=uptime(argv[1]);

	if (isatty(1)) {
	  char tmp[FMT_ULONG+20];
	  char tmp2[FMT_ULONG];
	  char* what;

	  if (pid==0) what="down "; else
	  if (pid==1) what="finished "; else {
	    len=fmt_str(tmp,"up (pid ");
	    len+=fmt_ulong(tmp+len,pid);
	    tmp[len+fmt_str(tmp+len,") ")]=0;
	    what=tmp;
	  }
	  tmp2[fmt_ulong(tmp2,ut)]=0;
	  msg(argv[1],": ",what,tmp2," seconds");
	} else {
	  char tmp[FMT_ULONG*2+5];
	  len=fmt_ulong(tmp,pid);
	  tmp[len]=' '; ++len;
	  len+=fmt_ulong(tmp+len,ut);
	  tmp[len]='\n';
	  write(1,tmp,len+1);
	}

	if (pid==0) return 2; else if (pid==1) return 3; else return 0;
      } else
	carp(argv[1],": no such service.");
      return 1;
    } else {
      int i;
      int ret=0;
      int sig=0;
      pid_t pid;
      if (argv[1][0]=='-') {
	switch (argv[1][1]) {
	case 'g':
	  for (i=2; i<argc; ++i) {
	    pid=__readpid(argv[i]);
	    if (pid<2) {
	      carp(argv[i],pid==1?": service terminated":": no such service");
	      ret=1;
	    } else {
	      char tmp[FMT_ULONG];
	      int i;
	      tmp[i=fmt_ulong(tmp,pid)]='\n';
	      write(1,tmp,i+1);
	    }
	  }
	  break;
	case 'p': sig=SIGSTOP; goto dokill; break;
	case 'c': sig=SIGCONT; goto dokill; break;
	case 'h': sig=SIGHUP; goto dokill; break;
	case 'a': sig=SIGALRM; goto dokill; break;
	case 'i': sig=SIGINT; goto dokill; break;
	case 't': sig=SIGTERM; goto dokill; break;
	case 'k': sig=SIGKILL; goto dokill; break;
	case 'o':
	  for (i=2; i<argc; ++i)
	    if (startservice(argv[i]) || respawn(argv[i],0)) {
	      carp("Could not start ",argv[i]);
	      ret=1;
	    }
	  break;
	case 'd':
	  for (i=2; i<argc; ++i) {
	    pid=__readpid(argv[i]);
	    if (pid==0) {
	      carp(argv[i],": no such service");
	      ret=1;
	    } else if (pid==1)
	      continue;
	    else
	      respawn(argv[i],0) || kill(pid,SIGTERM) || kill(pid,SIGCONT);
	  }
	  break;
	case 'u':
	  for (i=2; i<argc; ++i)
	    if (startservice(argv[i]) || respawn(argv[i],1)) {
	      carp("Could not start ",argv[i]);
	      ret=1;
	    }
	  break;
	case 'C':
	  for (i=2; i<argc; ++i)
	    if (check_remove(argv[i])) {
	      carp(argv[i]," could not be cleared");
	      ret=1;
	    }
	  break;
	case 'P':
	  pid=atoi(argv[1]+2);
	  if (pid>1)
	    if (setpid(argv[2],pid)) {
	      carp("Could not set PID of service ",argv[2]);
	      ret=1;
	    }
	  break;
	case 'H':
	  dumphistory();
	  break;
	case 'D':
	  dumpdependencies(argv[2]);
	  break;
	}
      }
      return ret;
dokill:
      for (i=2; i<argc; i++) {
	pid=__readpid(argv[i]);
	if (!pid) {
	  carp(argv[i],": no such service");
	  ret=1;
	} else if (pid==1) {
	  carp(argv[i],": service not running");
	  ret=1;
	} else if (kill(pid,sig)) {
	  char tmp[FMT_ULONG];
	  char tmp2[FMT_ULONG];
	  char* s;
	  switch (errno) {
	    case EINVAL: s="invalid signal"; break;
	    case EPERM: s="permission denied"; break;
	    case ESRCH: s="no such pid"; break;
	    default: s="unknown error";
	  }
	  tmp[fmt_ulong(tmp,sig)]=0;
	  tmp2[fmt_ulong(tmp2,pid)]=0;
	  carp(argv[i],": could not send signal ",tmp," to PID ",pid,": ",s);
	  ret=1;
	}
      }
      return ret;
    }
  } else {
    carp("minit: could not open " MINITROOT "/in or " MINITROOT "/out");
    return 1;
  }
}
